File: clients/macos/FastSortClientMac/Sources/FastSortClientMac/Services/Danmaku/Wechat/WechatNativeDanmakuAdapter.swift
import Foundation

struct WechatNativeRoomInit {
    let finderUsername: String
    let liveObjectId: String
    let liveId: String
    let description: String
    let liveCookies: String
}

@MainActor
final class WechatLiveAPIClient {
    private let sessionid: String
    private let wxuin: String
    private let aid = UUID().uuidString.lowercased()
    private let fingerprint = UUID().uuidString.replacingOccurrences(of: "-", with: "")

    private(set) var finderUsername = ""
    private(set) var liveObjectId = ""
    private(set) var liveId = ""
    private(set) var description = ""
    private(set) var liveCookies = ""
    private var lastStatus: Int?

    init(session: DanmakuWeChatSession, finderUsername: String? = nil) {
        self.sessionid = session.sessionid.replaceSpacesWithPlus()
        self.wxuin = session.wxuin
        self.finderUsername = finderUsername?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }

    func start() async throws -> WechatNativeRoomInit {
        if finderUsername.isEmpty {
            try await authData()
        } else {
            try? await authData()
        }
        try await helperUploadParams()
        try await checkLiveStatus(emitStatus: nil)
        guard !liveId.isEmpty, !liveObjectId.isEmpty else {
            throw NativeDanmakuAdapterError.notStarted("视频号")
        }
        try? await getLiveInfo()
        try await joinLive()
        try await onlineMember()
        return roomInit
    }

    func heartbeat() async throws -> NativeDanmakuStatus? {
        try await checkLiveStatus(emitStatus: nil)
        try? await onlineMember()
        guard let lastStatus else { return nil }
        return lastStatus == 1 ? .living : .stopped
    }

    func fetchMessages(requestRoomId: String?) async throws -> [NativeDanmakuEvent] {
        let response = try await post(
            path: "live/msg",
            referer: "https://channels.weixin.qq.com/platform/live/liveBuild",
            body: liveBody([
                "liveCookies": liveCookies,
                "longpollingScene": 0
            ])
        )
        let data = response["data"] as? [String: Any] ?? [:]
        if let nextLiveCookies = data["liveCookies"] as? String ?? data["live_cookies"] as? String,
           !nextLiveCookies.isEmpty {
            liveCookies = nextLiveCookies
        }
        var messageList = data["msgList"] as? [[String: Any]] ?? []
        if let respJsonStr = data["respJsonStr"] as? String,
           let innerData = respJsonStr.data(using: .utf8),
           let inner = try? JSONSerialization.jsonObject(with: innerData) as? [String: Any] {
            messageList.append(contentsOf: inner["msg_list"] as? [[String: Any]] ?? [])
        }
        return messageList.compactMap { decodeMessage($0, requestRoomId: requestRoomId) }
    }

    private var roomInit: WechatNativeRoomInit {
        WechatNativeRoomInit(
            finderUsername: finderUsername,
            liveObjectId: liveObjectId,
            liveId: liveId,
            description: description,
            liveCookies: liveCookies
        )
    }

    private func authData() async throws {
        let response = try await post(
            path: "auth/auth_data",
            referer: "https://channels.weixin.qq.com/platform/login-for-iframe?dark_mode=true&host_type=1",
            body: baseBody()
        )
        let data = response["data"] as? [String: Any] ?? [:]
        let finderUser = data["finderUser"] as? [String: Any] ?? [:]
        finderUsername = text(finderUser, "finderUsername")
        if finderUsername.isEmpty {
            throw NativeDanmakuAdapterError.loginExpired("视频号")
        }
    }

    private func helperUploadParams() async throws {
        _ = try await post(
            path: "helper/helper_upload_params",
            referer: "https://channels.weixin.qq.com/platform/login-for-iframe?dark_mode=true&host_type=1",
            body: baseBody(logFinderId: finderUsername)
        )
    }

    private func checkLiveStatus(emitStatus: ((NativeDanmakuStatus) -> Void)?) async throws {
        let response = try await post(
            path: "live/check_live_status",
            referer: "https://channels.weixin.qq.com/platform/live/home",
            body: baseBody(logFinderId: finderUsername)
        )
        let data = response["data"] as? [String: Any] ?? [:]
        liveId = text(data, "liveId")
        liveObjectId = text(data, "liveObjectId")
        description = text(data, "description")
        let status = NativeDanmakuHTTP.flexibleInt(data["status"])
        lastStatus = status
        if status == 1 {
            emitStatus?(.living)
        } else if status != nil {
            emitStatus?(.stopped)
        }
    }

    private func getLiveInfo() async throws {
        _ = try await post(
            path: "live/get_live_info",
            referer: "https://channels.weixin.qq.com/platform/live/liveBuild",
            body: liveInfoBody()
        )
    }

    private func joinLive() async throws {
        let response = try await post(
            path: "live/join_live",
            referer: "https://channels.weixin.qq.com/platform/live/liveBuild",
            body: liveBody([
                "timestamp": String(Int(Date().timeIntervalSince1970 * 1000))
            ])
        )
        let data = response["data"] as? [String: Any] ?? [:]
        liveCookies = text(data, "liveCookies")
        if liveCookies.isEmpty {
            throw NativeDanmakuAdapterError.notStarted("视频号")
        }
    }

    private func onlineMember() async throws {
        _ = try await post(
            path: "live/online_member",
            referer: "https://channels.weixin.qq.com/platform/live/liveBuild",
            body: liveBody([
                "clearRecentRewardHistory": true
            ])
        )
    }

    private func post(path: String, referer: String, body: [String: Any]) async throws -> [String: Any] {
        guard let url = requestURL(path: path, pageURL: referer) else {
            throw NativeDanmakuError("视频号接口 URL 构造失败：\(path)")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.httpShouldHandleCookies = false
        request.setValue("application/json, text/plain, */*", forHTTPHeaderField: "Accept")
        request.setValue("zh-CN,zh;q=0.9", forHTTPHeaderField: "Accept-Language")
        request.setValue("keep-alive", forHTTPHeaderField: "Connection")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("https://channels.weixin.qq.com", forHTTPHeaderField: "Origin")
        request.setValue(referer, forHTTPHeaderField: "Referer")
        request.setValue(NativeDanmakuHTTP.desktopUserAgent, forHTTPHeaderField: "User-Agent")
        request.setValue(wxuin.isEmpty ? "0000000000" : wxuin, forHTTPHeaderField: "X-WECHAT-UIN")
        request.setValue(fingerprint, forHTTPHeaderField: "finger-print-device-id")
        request.setValue("sessionid=\(sessionid); wxuin=\(wxuin)", forHTTPHeaderField: "Cookie")
        request.setValue("empty", forHTTPHeaderField: "Sec-Fetch-Dest")
        request.setValue("cors", forHTTPHeaderField: "Sec-Fetch-Mode")
        request.setValue("same-origin", forHTTPHeaderField: "Sec-Fetch-Site")
        request.setValue("\"Chromium\";v=\"136\", \"Google Chrome\";v=\"136\", \"Not.A/Brand\";v=\"99\"", forHTTPHeaderField: "sec-ch-ua")
        request.setValue("?0", forHTTPHeaderField: "sec-ch-ua-mobile")
        request.setValue("\"Windows\"", forHTTPHeaderField: "sec-ch-ua-platform")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw NativeDanmakuError("视频号接口请求失败：\(path) \(error.localizedDescription)")
        }
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw http.statusCode == 401 || http.statusCode == 403
                ? NativeDanmakuAdapterError.loginExpired("视频号")
                : NativeDanmakuError("视频号接口 \(path) HTTP \(http.statusCode)")
        }
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NativeDanmakuError("视频号接口 \(path) 返回不是 JSON")
        }
        let errCode = NativeDanmakuHTTP.flexibleInt(object["errCode"]) ?? 0
        if errCode == 300330 {
            throw NativeDanmakuAdapterError.loginExpired("视频号")
        }
        if errCode != 0 {
            let message = text(object, "errMsg")
            throw NativeDanmakuError(message.isEmpty ? "视频号接口 \(path) 错误：\(errCode)" : "视频号接口 \(path) 返回 \(errCode)：\(message)")
        }
        return object
    }

    private func requestURL(path: String, pageURL: String) -> URL? {
        var components = URLComponents(string: "https://channels.weixin.qq.com/cgi-bin/mmfinderassistant-bin/\(path)")
        components?.queryItems = [
            URLQueryItem(name: "_aid", value: aid),
            URLQueryItem(name: "_rid", value: randomRequestId()),
            URLQueryItem(name: "_pageUrl", value: pageURL)
        ]
        return components?.url
    }

    private func randomRequestId() -> String {
        let prefix = String(fingerprint.prefix(8))
        let suffix = UInt32.random(in: UInt32.min...UInt32.max)
        return "\(prefix)-\(String(format: "%08x", suffix))"
    }

    private func baseBody(logFinderId: String = "") -> [String: Any] {
        [
            "timestamp": String(Int(Date().timeIntervalSince1970 * 1000)),
            "_log_finder_uin": "",
            "_log_finder_id": logFinderId,
            "rawKeyBuff": NSNull(),
            "pluginSessionId": NSNull(),
            "scene": 7,
            "reqScene": 7
        ]
    }

    private func liveBody(_ extra: [String: Any]) -> [String: Any] {
        var body = baseBody(logFinderId: finderUsername)
        body["objectId"] = liveObjectId
        body["finderUsername"] = finderUsername
        body["liveId"] = liveId
        for (key, value) in extra {
            body[key] = value
        }
        return body
    }

    private func liveInfoBody() -> [String: Any] {
        var body = baseBody(logFinderId: finderUsername)
        body["liveObjectId"] = liveObjectId
        return body
    }

    private func decodeMessage(_ member: [String: Any], requestRoomId: String?) -> NativeDanmakuEvent? {
        guard NativeDanmakuHTTP.flexibleInt(member["type"]) == 1 else { return nil }
        let content = text(member, "content").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return nil }
        let seq = text(member, "seq")
        let clientMsgId = text(member, "clientMsgId").isEmpty ? text(member, "client_msg_id") : text(member, "clientMsgId")
        let messageId = seq.isEmpty ? (clientMsgId.isEmpty ? UUID().uuidString : clientMsgId) : seq
        let userId = decodedOpenId(from: clientMsgId) ?? text(member, "username")
        let userName = text(member, "nickname").isEmpty ? "视频号用户" : text(member, "nickname")
        return NativeDanmakuEvent(
            eventId: messageId,
            platform: "wechat",
            event: .chat,
            roomId: requestRoomId,
            platformRoomId: liveId,
            messageId: messageId,
            userId: userId,
            userName: userName,
            content: content,
            rawPayload: member
        )
    }

    private func decodedOpenId(from messageId: String) -> String? {
        guard let range = messageId.range(of: "_o9h") else { return nil }
        return String(messageId[messageId.index(after: range.lowerBound)...])
    }

    private func text(_ object: [String: Any], _ key: String) -> String {
        guard let value = object[key], !(value is NSNull) else { return "" }
        return "\(value)"
    }
}

@MainActor
final class WechatNativeDanmakuAdapter: NativeDanmakuAdapter {
    let platformKey = "wechat"
    let displayName = "视频号"

    private var preparedClientsByRoomKey: [String: WechatLiveAPIClient] = [:]
    private var preparedInitByRoomKey: [String: WechatNativeRoomInit] = [:]

    func prepare(_ request: NativeDanmakuConnectRequest) async throws -> NativeDanmakuConnectRequest {
        guard let session = DanmakuCookieSessionParser.wechatSession(fromLiveSession: request.liveSession) else {
            throw NativeDanmakuAdapterError.missingWeChatSession
        }
        let client = WechatLiveAPIClient(session: session, finderUsername: request.roomNumber)
        let roomInit = try await client.start()
        let key = cacheKey(for: request)
        preparedClientsByRoomKey[key] = client
        preparedInitByRoomKey[key] = roomInit
        return NativeDanmakuConnectRequest(
            platformKey: request.platformKey,
            roomId: request.roomId,
            roomNumber: roomInit.liveId,
            eid: roomInit.liveObjectId,
            liveType: request.liveType,
            liveSession: request.liveSession,
            cookieHeader: request.cookieHeader,
            displayName: request.displayName
        )
    }

    func connect(
        request: NativeDanmakuConnectRequest,
        onEvent: @escaping @MainActor (NativeDanmakuEvent) -> Void
    ) async throws -> NativeDanmakuConnection {
        let key = cacheKey(for: request)
        let client: WechatLiveAPIClient
        let roomInit: WechatNativeRoomInit
        if let preparedClient = preparedClientsByRoomKey.removeValue(forKey: key),
           let preparedInit = preparedInitByRoomKey.removeValue(forKey: key) {
            client = preparedClient
            roomInit = preparedInit
        } else {
            guard let session = DanmakuCookieSessionParser.wechatSession(fromLiveSession: request.liveSession) else {
                throw NativeDanmakuAdapterError.missingWeChatSession
            }
            client = WechatLiveAPIClient(session: session, finderUsername: request.roomNumber)
            roomInit = try await client.start()
        }

        let task = Task { @MainActor in
            onEvent(
                NativeDanmakuEvent(
                    platform: platformKey,
                    event: .status,
                    status: .living,
                    roomId: request.roomId,
                    platformRoomId: roomInit.liveId
                )
            )
            var lastHeartbeat = Date.distantPast
            while !Task.isCancelled {
                do {
                    if Date().timeIntervalSince(lastHeartbeat) >= 5 {
                        let status = try await client.heartbeat()
                        lastHeartbeat = Date()
                        if status == .stopped {
                            onEvent(NativeDanmakuEvent(platform: platformKey, event: .status, status: .stopped, roomId: request.roomId, platformRoomId: roomInit.liveId))
                            break
                        }
                    }
                    let events = try await client.fetchMessages(requestRoomId: request.roomId)
                    for event in events {
                        onEvent(event)
                    }
                    try await Task.sleep(nanoseconds: 300_000_000)
                } catch {
                    onEvent(
                        NativeDanmakuEvent(
                            platform: platformKey,
                            event: .error,
                            status: .error,
                            roomId: request.roomId,
                            platformRoomId: roomInit.liveId,
                            content: error.localizedDescription
                        )
                    )
                    break
                }
            }
            onEvent(NativeDanmakuEvent(platform: platformKey, event: .status, status: .disconnected, roomId: request.roomId, platformRoomId: roomInit.liveId))
        }

        return ClosureNativeDanmakuConnection(platformKey: platformKey) {
            task.cancel()
        }
    }

    private func cacheKey(for request: NativeDanmakuConnectRequest) -> String {
        request.roomId ?? request.displayName
    }
}

private extension String {
    func replaceSpacesWithPlus() -> String {
        replacingOccurrences(of: " ", with: "+")
    }
}

File: clients/windows/FastSort.Client.Windows/Core/Danmaku/Wechat/WechatNativeDanmakuAdapter.cs
using System.Net.Http;
using System.Text;
using System.Text.Json.Nodes;
using FastSort.Client.Windows.Core.Danmaku.Cookie;
using FastSort.Client.Windows.Core.Danmaku.Shared;

namespace FastSort.Client.Windows.Core.Danmaku.Wechat;

internal sealed record WechatNativeRoomInit(
    string FinderUsername,
    string LiveObjectId,
    string LiveId,
    string Description,
    string LiveCookies);

internal sealed class WechatLiveApiClient
{
    private static readonly HttpClient HttpClient = new();
    private readonly string _sessionId;
    private readonly string _wxUin;
    private readonly string _fingerprint = Guid.NewGuid().ToString("N");

    private string _finderUsername = "";
    private string _liveObjectId = "";
    private string _liveId = "";
    private string _description = "";
    private string _liveCookies = "";
    private int? _lastStatus;

    public WechatLiveApiClient(DanmakuWeChatSession session)
    {
        _sessionId = session.SessionId.Replace(" ", "+", StringComparison.Ordinal);
        _wxUin = session.WxUin;
    }

    public async Task<WechatNativeRoomInit> StartAsync(CancellationToken cancellationToken)
    {
        await AuthDataAsync(cancellationToken).ConfigureAwait(false);
        await HelperUploadParamsAsync(cancellationToken).ConfigureAwait(false);
        await CheckLiveStatusAsync(cancellationToken).ConfigureAwait(false);
        if (string.IsNullOrWhiteSpace(_liveId) || string.IsNullOrWhiteSpace(_liveObjectId))
        {
            throw new NativeDanmakuException("视频号当前账号未开播，未能解析 liveId/liveObjectId");
        }

        await GetLiveInfoAsync(cancellationToken).ConfigureAwait(false);
        await JoinLiveAsync(cancellationToken).ConfigureAwait(false);
        await OnlineMemberAsync(cancellationToken).ConfigureAwait(false);
        return RoomInit;
    }

    public async Task<NativeDanmakuStatus?> HeartbeatAsync(CancellationToken cancellationToken)
    {
        await CheckLiveStatusAsync(cancellationToken).ConfigureAwait(false);
        try { await OnlineMemberAsync(cancellationToken).ConfigureAwait(false); } catch { }
        return _lastStatus is null ? null : _lastStatus == 1 ? NativeDanmakuStatus.Living : NativeDanmakuStatus.Stopped;
    }

    public async Task<IReadOnlyList<NativeDanmakuEvent>> FetchMessagesAsync(string? requestRoomId, CancellationToken cancellationToken)
    {
        var response = await PostAsync(
            "live/msg",
            "https://channels.weixin.qq.com/platform/live/liveBuild",
            LiveBody(new JsonObject
            {
                ["liveCookies"] = _liveCookies,
                ["longpollingScene"] = 0
            }),
            cancellationToken).ConfigureAwait(false);

        var data = response["data"] as JsonObject ?? new JsonObject();
        var nextLiveCookies = NativeDanmakuHttp.FirstText(data, "liveCookies", "live_cookies");
        if (!string.IsNullOrWhiteSpace(nextLiveCookies))
        {
            _liveCookies = nextLiveCookies;
        }

        var messages = new List<JsonObject>();
        if (data["msgList"] is JsonArray list)
        {
            messages.AddRange(list.OfType<JsonObject>());
        }

        var respJsonStr = NativeDanmakuHttp.FirstText(data, "respJsonStr");
        if (!string.IsNullOrWhiteSpace(respJsonStr) &&
            NativeDanmakuHttp.ParseObject(respJsonStr) is { } inner &&
            inner["msg_list"] is JsonArray innerList)
        {
            messages.AddRange(innerList.OfType<JsonObject>());
        }

        return messages
            .Select(message => DecodeMessage(message, requestRoomId))
            .Where(evt => evt is not null)
            .Select(evt => evt!)
            .ToList();
    }

    private WechatNativeRoomInit RoomInit => new(_finderUsername, _liveObjectId, _liveId, _description, _liveCookies);

    private async Task AuthDataAsync(CancellationToken cancellationToken)
    {
        var response = await PostAsync(
            "auth/auth_data",
            "https://channels.weixin.qq.com/platform/login-for-iframe?dark_mode=true&host_type=1",
            BaseBody(),
            cancellationToken).ConfigureAwait(false);

        var data = response["data"] as JsonObject ?? new JsonObject();
        var finderUser = data["finderUser"] as JsonObject ?? new JsonObject();
        _finderUsername = NativeDanmakuHttp.FirstText(finderUser, "finderUsername");
        if (string.IsNullOrWhiteSpace(_finderUsername))
        {
            throw new NativeDanmakuException("视频号登录态已失效，请重新扫码登录视频号工作台");
        }
    }

    private Task HelperUploadParamsAsync(CancellationToken cancellationToken)
    {
        return PostAsync(
            "helper/helper_upload_params",
            "https://channels.weixin.qq.com/platform/login-for-iframe?dark_mode=true&host_type=1",
            BaseBody(_finderUsername),
            cancellationToken);
    }

    private async Task CheckLiveStatusAsync(CancellationToken cancellationToken)
    {
        var response = await PostAsync(
            "live/check_live_status",
            "https://channels.weixin.qq.com/platform/live/home",
            BaseBody(_finderUsername),
            cancellationToken).ConfigureAwait(false);

        var data = response["data"] as JsonObject ?? new JsonObject();
        _liveId = NativeDanmakuHttp.FirstText(data, "liveId");
        _liveObjectId = NativeDanmakuHttp.FirstText(data, "liveObjectId");
        _description = NativeDanmakuHttp.FirstText(data, "description");
        _lastStatus = NativeDanmakuHttp.FlexibleInt(data["status"]);
    }

    private Task GetLiveInfoAsync(CancellationToken cancellationToken)
    {
        return PostAsync(
            "live/get_live_info",
            "https://channels.weixin.qq.com/platform/live/liveBuild",
            LiveBody(new JsonObject()),
            cancellationToken);
    }

    private async Task JoinLiveAsync(CancellationToken cancellationToken)
    {
        var response = await PostAsync(
            "live/join_live",
            "https://channels.weixin.qq.com/platform/live/liveBuild",
            LiveBody(new JsonObject
            {
                ["timestamp"] = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString()
            }),
            cancellationToken).ConfigureAwait(false);

        var data = response["data"] as JsonObject ?? new JsonObject();
        _liveCookies = NativeDanmakuHttp.FirstText(data, "liveCookies");
        if (string.IsNullOrWhiteSpace(_liveCookies))
        {
            throw new NativeDanmakuException("视频号当前账号未开播，join_live 未返回 liveCookies");
        }
    }

    private Task OnlineMemberAsync(CancellationToken cancellationToken)
    {
        return PostAsync(
            "live/online_member",
            "https://channels.weixin.qq.com/platform/live/liveBuild",
            LiveBody(new JsonObject
            {
                ["clearRecentRewardHistory"] = true
            }),
            cancellationToken);
    }

    private async Task<JsonObject> PostAsync(string path, string referer, JsonObject body, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, $"https://channels.weixin.qq.com/cgi-bin/mmfinderassistant-bin/{path}");
        request.Headers.TryAddWithoutValidation("Accept", "application/json, text/plain, */*");
        request.Headers.TryAddWithoutValidation("Accept-Language", "zh-CN,zh;q=0.9");
        request.Headers.TryAddWithoutValidation("Origin", "https://channels.weixin.qq.com");
        request.Headers.TryAddWithoutValidation("Referer", referer);
        request.Headers.TryAddWithoutValidation("User-Agent", NativeDanmakuHttp.DesktopUserAgent);
        request.Headers.TryAddWithoutValidation("X-WECHAT-UIN", string.IsNullOrWhiteSpace(_wxUin) ? "0000000000" : _wxUin);
        request.Headers.TryAddWithoutValidation("finger-print-device-id", _fingerprint);
        request.Headers.TryAddWithoutValidation("Cookie", $"sessionid={_sessionId}; wxuin={_wxUin}");
        request.Content = new StringContent(body.ToJsonString(NativeDanmakuHttp.JsonOptions), Encoding.UTF8, "application/json");

        using var response = await HttpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new NativeDanmakuException((int)response.StatusCode is 401 or 403
                ? "视频号登录态已失效，请重新扫码登录视频号工作台"
                : $"视频号接口 HTTP {(int)response.StatusCode}");
        }

        var root = NativeDanmakuHttp.ParseObject(await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false))
            ?? throw new NativeDanmakuException("视频号接口返回不是 JSON");
        var errCode = NativeDanmakuHttp.FlexibleInt(root["errCode"]) ?? 0;
        if (errCode == 300330)
        {
            throw new NativeDanmakuException("视频号登录态已失效，请重新扫码登录视频号工作台");
        }

        if (errCode != 0)
        {
            var message = NativeDanmakuHttp.FirstText(root, "errMsg");
            throw new NativeDanmakuException(string.IsNullOrWhiteSpace(message) ? $"视频号接口错误：{errCode}" : message);
        }

        return root;
    }

    private static JsonObject BaseBody(string logFinderId = "")
    {
        return new JsonObject
        {
            ["timestamp"] = DateTimeOffset.Now.ToUnixTimeMilliseconds().ToString(),
            ["_log_finder_uin"] = "",
            ["_log_finder_id"] = logFinderId,
            ["rawKeyBuff"] = null,
            ["pluginSessionId"] = null,
            ["scene"] = 7,
            ["reqScene"] = 7
        };
    }

    private JsonObject LiveBody(JsonObject extra)
    {
        var body = BaseBody(_finderUsername);
        body["objectId"] = _liveObjectId;
        body["finderUsername"] = _finderUsername;
        body["liveId"] = _liveId;
        foreach (var item in extra)
        {
            body[item.Key] = item.Value?.DeepClone();
        }

        return body;
    }

    private NativeDanmakuEvent? DecodeMessage(JsonObject member, string? requestRoomId)
    {
        if (NativeDanmakuHttp.FlexibleInt(member["type"]) != 1)
        {
            return null;
        }

        var content = NativeDanmakuHttp.FirstText(member, "content").Trim();
        if (string.IsNullOrWhiteSpace(content))
        {
            return null;
        }

        var seq = NativeDanmakuHttp.FirstText(member, "seq");
        var clientMsgId = NativeDanmakuHttp.FirstText(member, "clientMsgId", "client_msg_id");
        var messageId = !string.IsNullOrWhiteSpace(seq) ? seq : !string.IsNullOrWhiteSpace(clientMsgId) ? clientMsgId : Guid.NewGuid().ToString("N");
        var userId = DecodedOpenId(clientMsgId) ?? NativeDanmakuHttp.FirstText(member, "username");
        var userName = NativeDanmakuHttp.FirstText(member, "nickname");
        if (string.IsNullOrWhiteSpace(userName))
        {
            userName = "视频号用户";
        }

        return new NativeDanmakuEvent
        {
            EventId = messageId,
            Platform = "wechat",
            Event = NativeDanmakuEventKind.Chat,
            RoomId = requestRoomId,
            PlatformRoomId = _liveId,
            MessageId = messageId,
            UserId = userId,
            UserName = userName,
            Content = content,
            RawPayload = new JsonObject
            {
                ["wxMsgId"] = messageId,
                ["danmuUserId"] = userId,
                ["danmuUserName"] = userName,
                ["danmuContent"] = content,
                ["wxRoomId"] = _liveId
            }.ToJsonString(NativeDanmakuHttp.JsonOptions)
        };
    }

    private static string? DecodedOpenId(string messageId)
    {
        var index = messageId.IndexOf("_o9h", StringComparison.Ordinal);
        return index < 0 || index + 1 >= messageId.Length ? null : messageId[(index + 1)..];
    }
}

public sealed class WechatNativeDanmakuAdapter : INativeDanmakuAdapter
{
    public string PlatformKey => "wechat";

    public async Task<INativeDanmakuConnection> ConnectAsync(
        NativeDanmakuConnectRequest request,
        Func<NativeDanmakuEvent, Task> onEvent,
        CancellationToken cancellationToken)
    {
        try
        {
            var session = DanmakuCookieSessionParser.WeChatSessionFromLiveSession(request.LiveSession);
            if (session is null)
            {
                throw new NativeDanmakuException("视频号 Cookie 中缺少 sessionid 或 wxuin，请重新扫码登录视频号工作台");
            }

            var client = new WechatLiveApiClient(session);
            var roomInit = await client.StartAsync(cancellationToken).ConfigureAwait(false);
            await onEvent(NativeDanmakuEvent.StatusEvent(PlatformKey, NativeDanmakuStatus.Living, $"已进入视频号直播：{roomInit.LiveId}")).ConfigureAwait(false);

            var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            var loop = Task.Run(async () =>
            {
                var lastHeartbeat = DateTimeOffset.MinValue;
                while (!cts.IsCancellationRequested)
                {
                    try
                    {
                        if (DateTimeOffset.Now - lastHeartbeat >= TimeSpan.FromSeconds(5))
                        {
                            var status = await client.HeartbeatAsync(cts.Token).ConfigureAwait(false);
                            lastHeartbeat = DateTimeOffset.Now;
                            if (status == NativeDanmakuStatus.Stopped)
                            {
                                await onEvent(NativeDanmakuEvent.StatusEvent(PlatformKey, NativeDanmakuStatus.Stopped, "视频号直播已结束")).ConfigureAwait(false);
                                break;
                            }
                        }

                        var events = await client.FetchMessagesAsync(request.RoomId, cts.Token).ConfigureAwait(false);
                        foreach (var evt in events)
                        {
                            await onEvent(evt).ConfigureAwait(false);
                        }

                        await Task.Delay(300, cts.Token).ConfigureAwait(false);
                    }
                    catch (OperationCanceledException)
                    {
                        break;
                    }
                    catch (Exception ex)
                    {
                        await onEvent(NativeDanmakuEvent.ErrorEvent(PlatformKey, ex.Message)).ConfigureAwait(false);
                        break;
                    }
                }

                await onEvent(NativeDanmakuEvent.StatusEvent(PlatformKey, NativeDanmakuStatus.Disconnected, "视频号 native adapter 已断开")).ConfigureAwait(false);
            }, cts.Token);

            return new NativeDanmakuConnection(PlatformKey, NativeDanmakuStatus.Living, async _ =>
            {
                await cts.CancelAsync().ConfigureAwait(false);
                try { await loop.ConfigureAwait(false); } catch (OperationCanceledException) { }
                cts.Dispose();
            });
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            await onEvent(NativeDanmakuEvent.ErrorEvent(PlatformKey, ex.Message)).ConfigureAwait(false);
            return new NativeDanmakuConnection(PlatformKey, NativeDanmakuStatus.Error);
        }
    }
}
